### ISSUE TITLE
Initial Reference Implementation of Decentralized Authentication (DID-Auth) and Authorization Mechanisms

### ISSUE THREAD

**Issue by ccoldwell** [bcgov/TheOrgBook#165]
Value: $50,000.00Closes: Wednesday, January 31, 2018Location: Victoria In-person work NOT required<h2>Opportunity Description</h2><p dir="ltr" style="line-height: 1.38; margin-top: 18pt; margin-bottom: 6pt;"><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;">One of the key promises of </span><a style="text-decoration: none;" href="https://w3c-ccg.github.io/did-spec/"><span style="font-size: 11pt; font-family: Arial; color: #1155cc; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: underline; -webkit-text-decoration-skip: none; text-decoration-skip-ink: none; vertical-align: baseline; white-space: pre-wrap;">decentralized identifiers (DIDs)</span></a><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;"> and </span><a style="text-decoration: none;" href="https://bitsonblocks.net/2017/05/17/a-gentle-introduction-to-self-sovereign-identity/"><span style="font-size: 11pt; font-family: Arial; color: #1155cc; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: underline; -webkit-text-decoration-skip: none; text-decoration-skip-ink: none; vertical-align: baseline; white-space: pre-wrap;">self-sovereign identity (SSI)</span></a><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;"> is the potential for eliminating password-based authentication, particularly on websites, and the resulting improvement in online security. The human element tends to be the weakest point of any online security system and, to put it bluntly, passwords are the worst. As such, improving the human components of the online security infrastructure will have the greatest impact on improving overall online security for our citizens.</span></p>
<p><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;">A working group in the </span><a style="text-decoration: none;" href="http://identity.foundation/"><span style="font-size: 11pt; font-family: Arial; color: #1155cc; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: underline; -webkit-text-decoration-skip: none; text-decoration-skip-ink: none; vertical-align: baseline; white-space: pre-wrap;">Decentralized Identity Foundation (DIF)</span></a><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;"> has been discussing the use of DIDs for authentication. A </span><a style="text-decoration: none;" href="https://docs.google.com/document/d/1Lt0uMvSuv094Bb-5XvVKNqNFEDrlWm3acy1O5-vVZu4/edit"><span style="font-size: 11pt; font-family: Arial; color: #1155cc; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: underline; -webkit-text-decoration-skip: none; text-decoration-skip-ink: none; vertical-align: baseline; white-space: pre-wrap;">preliminary concept paper</span></a><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;"> has been created stemming from discussions at the </span><a style="text-decoration: none;" href="https://github.com/swcurran/rebooting-the-web-of-trust-spring2017/blob/fc8a7861c82043edb6e284497917d6804bbc7023/event-documents/group-abstracts/did-auth.md"><span style="font-size: 11pt; font-family: Arial; color: #1155cc; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: underline; -webkit-text-decoration-skip: none; text-decoration-skip-ink: none; vertical-align: baseline; white-space: pre-wrap;">Rebooting the Web of Trust 2017 Conference</span></a><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;"> as the starting point for the discussion amongst the DIF group. It appears based on the ideas outlined by the members of the DIF group, progress on the DID Spec and on the </span><a style="text-decoration: none;" href="https://medium.com/decentralized-identity/a-universal-resolver-for-self-sovereign-identifiers-48e6b4a5cc3c"><span style="font-size: 11pt; font-family: Arial; color: #1155cc; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: underline; -webkit-text-decoration-skip: none; text-decoration-skip-ink: none; vertical-align: baseline; white-space: pre-wrap;">Universal Resolver for DIDs</span></a><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;">, we are at a point where a working implementation of DID-Auth would be useful.</span></p>
<p><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;">We propose an implementation of Decentralized Identifier Authentication (DID-Auth) which we can use with TheOrgBook (</span><a style="text-decoration: none;" href="https://bcgov.github.io/TheOrgBook/"><span style="font-size: 11pt; font-family: Arial; color: #1155cc; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: underline; -webkit-text-decoration-skip: none; text-decoration-skip-ink: none; vertical-align: baseline; white-space: pre-wrap;">described here</span></a><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;">, </span><a style="text-decoration: none;" href="https://devex-von-test.pathfinder.gov.bc.ca/home"><span style="font-size: 11pt; font-family: Arial; color: #1155cc; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: underline; -webkit-text-decoration-skip: none; text-decoration-skip-ink: none; vertical-align: baseline; white-space: pre-wrap;">development instance here</span></a><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;">) website. The current and primary features of TheOrgBook are public, including searching for/discovery of organizations, displaying organizational informational, and viewing the details of an organization's verifiable claims. We also propose some initial work on Authorization using Verifiable Claims. There are several use cases related to TheOrgBook for DID-Auth, and Verifiable Claims Authorization:</span></p>
<p><span style="text-decoration: underline; font-size: 11pt;"><span style="font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-variant-east-asian: normal; font-variant-position: normal; text-decoration: underline; vertical-align: baseline; white-space: pre-wrap;">Authentication Scenarios</span></span></p>
<ol>
<li><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: bold; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;">System to System:</span><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;"> Services that generate verifiable claims about organizations and write them to TheOrgBook need to be authenticated. Such Services must have a DID known to TheOrgBook and use that DID to access the Issuer API for writing verifiable claims to TheOrgBook. This would be a generic authentication method between two services at the API level.</span></li>
<li><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: bold; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;">Administrative:</span><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;"> There is an administrative element of TheOrgBook that should be limited to authorized users who would have a DID.</span></li>
</ol>
<p><span style="text-decoration: underline; font-size: 11pt;"><span style="font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-variant-east-asian: normal; font-variant-position: normal; text-decoration: underline; vertical-align: baseline; white-space: pre-wrap;">Authorization Scenarios</span></span></p>
<ol>
<li dir="ltr" style="line-height: 1.38; margin-top: 0pt; margin-bottom: 0pt;"><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: bold; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;">Claim Your Claims:</span><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;"> There is a long term desire to support verifying Organization Owners using some (likely offline) process, providing them with a Verifiable Claim that enables them to "Claim your Claims" on TheOrgBook, and thus, giving them the ability to extend their organization's TheOrgBook page (add accreditations such as the BBB, ratings, product / service information, etc.).</span></li>
<li dir="ltr" style="line-height: 1.38; margin-top: 0pt; margin-bottom: 0pt;"><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: bold; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;">Delegation:</span><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;"> Further, since an Organization's owner is not likely the only person associated with an organization that will need access to TheOrgBook, there will be a need to extend the authentication process to support organizational owners delegating access to their TheOrgBook page as they see fit.</span></li>
</ol>
<p dir="ltr" style="line-height: 1.38; margin-top: 0pt; margin-bottom: 0pt;"><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;">We would like to explore implementing at least the first three of these scenarios using DID-Auth, and ideally all four.<br /><br /></span></p>
<p dir="ltr" style="line-height: 1.38; margin-top: 0pt; margin-bottom: 0pt;"><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;">For the system to system and administrative use cases, no verifiable claims are necessary for authorizations - the API and website manage authorizations of users granted access to specific capabilities. The website would maintain a user table of registered DIDs that are permitted access to the write verifiable claims and administrative features of the TheOrgBook, and a process is executed to authenticate access requests to the site. The administrative implementation must consider performance and usability for both the initial authentication process (is the process as fast/easy as passwords?) and ongoing verification (e.g. session renewal/expiration).<br /><br /></span></p>
<p dir="ltr" style="line-height: 1.38; margin-top: 0pt; margin-bottom: 0pt;"><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;">The "Claim your Claims" use case extends the administrative use case by adding authorization via Verifiable Claims. DID-Auth will be executed and on success, a Verifiable Claim proof process executed to determine the resources (Organizational Page and related data) to which the user will be granted access. Again, performance and usability of the process will be paramount - the user must find the process no harder than using traditional approaches. Further, the overhead on the website side - the limited per user information to be maintained - will be of interest. A goal will be demonstrating how websites can operate effectively without collecting and maintaining private information about users. </span></p>
<p>The delegation of authority use case extends the "Claim your Claims" authorization functionality by providing the Organizational Owner with the ability to delegate their access to others under their control - e.g. without having to log into TheOrgBook to record the delegation or to revoke that access. Although this use case could get into a lot of client side functionality - for example, how the Owner manages their delegations - our focus will be on the DID-Auth part of the of process and assume that the client side challenges will be handled by others.</p>
<p><span style="text-decoration: underline;">Additional Considerations</span></p>
<p dir="ltr" style="line-height: 1.38; margin-top: 0pt; margin-bottom: 0pt;"><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;">While we know that the implementation will not be the last word in creating DID-Auth, but an initial set of steps. We would like the following considered for this implementation:</span></p>
<ul>
<li dir="ltr" style="line-height: 1.38; margin-top: 0pt; margin-bottom: 0pt;"><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;">The flows described in the docs linked in the introduction match what we are assuming will be used here. While these will be the starting point, we expect some collaboration with the community to evolve these flows, balancing the need for creating working code against the goal of creating an &ldquo;ideal implementation&rdquo;.<br /></span></li>
<li dir="ltr" style="line-height: 1.38; margin-top: 0pt; margin-bottom: 0pt;"><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;">The Universal Resolver will ideally be used in going from a DID to the authentication services associated with that DID.</span></li>
<li dir="ltr" style="line-height: 1.38; margin-top: 0pt; margin-bottom: 0pt;"><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;">Ideally, a Universal Authenticator capability would be an outcome of this effort, allowing (at least for non-Verifiable Claim-based authentication) execution work across DID method implementations.</span></li>
<li dir="ltr" style="line-height: 1.38; margin-top: 0pt; margin-bottom: 0pt;"><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;">The API/website functionality should have a minimum impact to current authentication mechanisms so as to reduce adoption effort. Ideally, the solution would be a compatible enhancement to commonly used authentication mechanisms. For example, is there a common JS library used for authentication that could be built upon to add DID-Auth functionality?</span>
<ul>
<li dir="ltr" style="line-height: 1.38; margin-top: 0pt; margin-bottom: 0pt;"><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;">Note: This is just a consideration for this effort. There will not be a requirement for submitting an update to an existing library/component for this need.</span></li>
</ul>
</li>
</ul>
<p><span style="text-decoration: underline; font-size: 11pt;"><span style="font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-variant-east-asian: normal; font-variant-position: normal; text-decoration: underline; vertical-align: baseline; white-space: pre-wrap;">Assumptions</span></span></p>
<ul>
<li dir="ltr" style="line-height: 1.38; margin-top: 0pt; margin-bottom: 0pt;"><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;">The implementation need not be concerned with no-Auth-related capabilities of the components. For example, although some sort of DID-capable client will be needed for demonstrating the login process, issues like general usability, scalability, and key management are not a part of this effort.</span></li>
<li dir="ltr" style="line-height: 1.38; margin-top: 0pt; margin-bottom: 0pt;"><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;">We have full control over the code and deployment of TheOrgBook and so can be expedient in making changes to the application as necessary to support the work defined here.</span></li>
<li dir="ltr" style="line-height: 1.38; margin-top: 0pt; margin-bottom: 0pt;"><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;">Nice to have, but not required, is an implementation of an effective &ldquo;old and new&rdquo; user experience. That is, the creation of a user experience that supports both password-style and DID-Auth login to the site. It is sufficient for the purposes of this implementation that TheOrgBook support only DID-Auth.</span></li>
</ul>
<p><span style="text-decoration: underline; font-size: 11pt;"><span style="font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-variant-east-asian: normal; font-variant-position: normal; text-decoration: underline; vertical-align: baseline; white-space: pre-wrap;">Opportunity</span></span></p>
<p>This is an opportunity to work with the Verifiable Organizations Network team to deliver enhancements to both a BC Government project and potentialyl an open standard.</p>
<p><span style="font-size: 10.5pt; font-family: Arial; color: #333333; background-color: transparent; font-variant-ligatures: normal; font-variant-east-asian: normal; font-variant-position: normal; vertical-align: baseline; white-space: pre-wrap;">The fixed-price reward is for a potential total of </span><span style="font-size: 10.5pt; font-family: Arial; color: #333333; background-color: transparent; font-weight: bold; font-variant-ligatures: normal; font-variant-east-asian: normal; font-variant-position: normal; vertical-align: baseline; white-space: pre-wrap;">$CDN 50,000*</span><span style="font-size: 10.5pt; font-family: Arial; color: #333333; background-color: transparent; font-variant-ligatures: normal; font-variant-east-asian: normal; font-variant-position: normal; vertical-align: baseline; white-space: pre-wrap;"> for satisfaction of the </span><span style="font-size: 10.5pt; font-family: Arial; color: #333333; background-color: transparent; font-weight: bold; font-variant-ligatures: normal; font-variant-east-asian: normal; font-variant-position: normal; vertical-align: baseline; white-space: pre-wrap;">Acceptance Criteria</span><span style="font-size: 10.5pt; font-family: Arial; color: #333333; background-color: transparent; font-variant-ligatures: normal; font-variant-east-asian: normal; font-variant-position: normal; vertical-align: baseline; white-space: pre-wrap;"> below, per this payment schedule:</span></p>
<ul>
<li>Completion of Phase 1 - $10,000</li>
<li dir="ltr" style="line-height: 1.38; margin-top: 0pt; margin-bottom: 0pt;"><span style="font-size: 10.5pt; font-family: Arial; color: #333333; background-color: transparent; font-variant-ligatures: normal; font-variant-east-asian: normal; font-variant-position: normal; vertical-align: baseline; white-space: pre-wrap;">Completion of Phase 2 - $5,000</span></li>
<li dir="ltr" style="line-height: 1.38; margin-top: 0pt; margin-bottom: 0pt;"><span style="font-size: 10.5pt; font-family: Arial; color: #333333; background-color: transparent; font-variant-ligatures: normal; font-variant-east-asian: normal; font-variant-position: normal; vertical-align: baseline; white-space: pre-wrap;">Completion of Phase 3 - $10,000</span></li>
<li dir="ltr" style="line-height: 1.38; margin-top: 0pt; margin-bottom: 0pt;"><span style="font-size: 10.5pt; font-family: Arial; color: #333333; background-color: transparent; font-variant-ligatures: normal; font-variant-east-asian: normal; font-variant-position: normal; vertical-align: baseline; white-space: pre-wrap;">Completion of Phase 4 - $5,000</span></li>
<li dir="ltr" style="line-height: 1.38; margin-top: 0pt; margin-bottom: 0pt;"><span style="font-size: 10.5pt; font-family: Arial; color: #333333; background-color: transparent; font-variant-ligatures: normal; font-variant-east-asian: normal; font-variant-position: normal; vertical-align: baseline; white-space: pre-wrap;">Completion of Phase 5 - $10,000</span></li>
<li dir="ltr" style="line-height: 1.38; margin-top: 0pt; margin-bottom: 8pt;"><span style="font-size: 10.5pt; font-family: Arial; color: #333333; background-color: transparent; font-variant-ligatures: normal; font-variant-east-asian: normal; font-variant-position: normal; vertical-align: baseline; white-space: pre-wrap;">Completion of Phase 6 - $10,000</span></li>
</ul>
<p><span style="font-size: 10.5pt; font-family: Arial; color: #333333; background-color: transparent; font-variant-ligatures: normal; font-variant-east-asian: normal; font-variant-position: normal; vertical-align: baseline; white-space: pre-wrap;">* The full amount is dependent on an evaluation of outputs at the end of Phase 1, during which </span><span style="font-size: 11pt; font-family: Arial; background-color: transparent; font-variant-ligatures: normal; font-variant-east-asian: normal; font-variant-position: normal; vertical-align: baseline; white-space: pre-wrap;">Verifiable Organizations Network team</span><span style="font-size: 10.5pt; font-family: Arial; color: #333333; background-color: transparent; font-variant-ligatures: normal; font-variant-east-asian: normal; font-variant-position: normal; vertical-align: baseline; white-space: pre-wrap;"> will determine whether or not work will continue into phases 2 and 3.</span></p><h2>Acceptance Criteria</h2><ol>
<li><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;">Agreed upon technical approach, implementation components, and interaction flow diagrams for the authentication only use cases - System to System and Administrative. This work needs to be shared with relevant communities with could include W3C, Decentralized Identity Foundation, Rebooting the Web of Trust or others as appropriate.</span></li>
<li><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;">Working code implementing the System to System use case to protect the TheOrgBook API endpoints related to Services issuing claims to TheOrgBook. Requires code for both TheOrgBook and the VON-Connector code that Services use for issuing claims to TheOrgBook.</span></li>
<li><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;">Working code implementing the Administrative use case supporting a browser-based user logging into a web session on TheOrgBook as an identified user. Requires code for both TheOrgBook and a Web App accessing TheOrgBook.</span></li>
<li><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;">An agreed upon technical approach and interaction flow diagrams extending the authentication only use cases to include authorization via Verifiable Claims. This work needs to be shared with relevant communities with could include W3C, Decentralized Identity Foundation, Rebooting the Web of Trust or others as appropriate.</span></li>
<li><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;">Working code implementing the &ldquo;Claim your Claims&rdquo; use case supporting a browser-based user logging into a web session with access to specific resources based on a Verifiable Claim held by the user. Requires code for both TheOrgBook and a Web App accessing TheOrgBook.</span></li>
<li><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;">A final write up summarizing the work done, including two key elements:</span>
<ol>
<li><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;">Post-implementation recommendations for changes/improvements to the delivered code (e.g. &ldquo;what could we have done better?&rdquo;).</span></li>
<li><span style="font-size: 11pt; font-family: Arial; color: #000000; background-color: transparent; font-weight: 400; font-style: normal; font-variant: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;">A proposed technical approach and interaction flow diagrams for the Delegation Use Case.</span></li>
</ol>
</li>
</ol><h2>How to Apply</h2><p>Go to the <a href="https://bcdevexchange.org/opportunities/opp-initial-reference-implementation-of-decentralized-authentication--did-auth--and-authorization-mechanisms">Opportunity Page</a>, click the Apply button above and submit your proposal by 16:00 PST on Wednesday, January 31, 2018</b>.</p><p>We plan to assign this opportunity by <b>Friday, February 2, 2018</b> with work to start on <b>Monday, February 5, 2018</b>.</p><h2>Proposal Evaluation Criteria</h2><ol>
<li>Your approach to completing the Acceptance Criteria in a short proposal (1-2 pages) which includes evidence to support the criteria outlined in items 2-5 below. Evidence can include GitHub IDs, projects for example. (20 points)</li>
<li>Your prior experience with open source projects in identity, self-sovereign identity, verifiable claims, decentralized identifiers, distributed ledger/blockchain or other relevant technologies. (30 points)</li>
<li>Your expertise in authentication and authorization protocol design and implementation. (30 points)</li>
<li>Your experience in contributing to open standards and/or open source communities (20 points)</li>
<li>Your ability to satisfy the Acceptance Criteria on or before 31 March 2018 (10 points)</li>
</ol>

**Comment by jljordan42** [comment]
Hi Everyone ... happy to answers any questions that arise from this opportunity.

**Comment by mojitoholic** [comment]
Quick clarification about the timeline: Does item 5 in the evaluation criteria mean that all six phases should be completed before the end of March 2018, or only Phase 1? In other words, are you envisioning the entire project to be completed in 2 months?

**Comment by jljordan42** [comment]
Thanks for your question. 
The project must be completed by March 31 2018. 
Best
John

**Comment by jljordan42** [comment]
Thanks to everyone for their interesting and bids. We are pleased to assign this opportunity to Markus (peacekeeper). If you bid and would like a debrief .. please contact me.

**Comment by ccoldwell** [comment]
This opportunity has been assigned

**Comment by jljordan42** [comment]
The work has been completed and accepted.
https://github.com/topics/did-auth are the resulting repositories.
